### ISSUE TITLE
Don't auto-expand large documents

### ISSUE THREAD

**Issue by drudge** [bobthecow/genghis#12]
I've noticed that with some collections I have which contain either a large number of properties or fairly large embedded arrays, genghis will pretty much freeze up for sometimes up to a minute while it renders the JSON. It would be nice if there was a way to limit the number of expanded items, or disabled auto expanding completely.

**Comment by bobthecow** [comment]
I agree. I've been working on a good answer to this problem for a while. Some recent changes (collapsing documents in a collection by default if there are more than 10 visible) have helped, but we're not there yet.

I'd prefer to solve this without resorting to preferences. Genghis should be smart enough not to do anything automatically which will make itself sluggish. This most likely means that large arrays in documents will be collapsed by default, and wide or deep documents will be collapsed by default.

Another big piece of this is the sheer number of DOM elements generated when rendering a large document. I've started rewriting the JSON printer — twice actually — to drastically reduce the amount of markup needed to render a document. The new JSON printer is definitely far more performant, based on that change alone.

I hope to have improvements, if not the final solution, in the next release.

**Comment by bobthecow** [comment]
Incidentally, from my testing, it seems to be displaying the JSON that freezes Genghis, not rendering it.

**Comment by crueber** [comment]
I was just thinking this, as I was trying it out. I have some documents that get pretty darn long. Here's one suggestion- Perhaps if the number of lines taken up by the document is longer than X (where X is some sane number), it automatically code-folds it?

Just one thought!

**Comment by bobthecow** [comment]
@crueber I'm leaning towards auto-collapsing based on document height to window height ratio... mebbe if it's 2x or 3x the height of the window, automatically collapse it?

**Comment by crueber** [comment]
@bobthecow Actually, I really like that idea. It would predictably collapse documents that are too large for the screen, rather than the data. That tends to be what I care about. That also limits the initial height of the page (on load) to something fairly manageable. 2x the window seems like it should be plenty, though complex applications may easily hit that with regularity, whereas 3x might not be hit often at all. Personally, I'd tend towards 2x because it limits the initial viewport. The only trade offs I can think of are pretty well mitigated by having a nice sane default.

If you think that's a good direction, I wouldn't mind implementing this. Interested?

**Comment by bobthecow** [comment]
@crueber I always welcome a pull request :)

**Comment by mavencode01** [comment]
Is there anyway to add a preview on reference document without actually clicking on the DBRef link. I think this will be nice to have especially when you have large reference dataset.

**Comment by alexturpin** [comment]
If someone isn't working on implementing this I'd be down to give it a try once my schedule frees up a little more.

**Comment by crueber** [comment]
I didn't ever pick it up.
